File: platform/winui/LUI.Core/LUIExtensionRegistry.cs
// Port of platform/flutter/lib/lui_flutter_extension.dart: the extension
// schema registry. Rendering-side builder delegates live in LUI.WinUI;
// this registry only tracks the schema each `create-extension` op must
// match by identifier and fingerprint.

using System.Collections.Generic;
using System.Text.RegularExpressions;

namespace LUI
{
    public enum LUIExtensionValueKind
    {
        String,
        Boolean,
        Integer,
        DoubleValue,
    }

    public static class LUIExtensionValueKindExtensions
    {
        public static bool Accepts(
            this LUIExtensionValueKind kind, LUIWireValue value) =>
            kind switch
            {
                LUIExtensionValueKind.String => value is LUIWireValue.String,
                LUIExtensionValueKind.Boolean => value is LUIWireValue.Bool,
                LUIExtensionValueKind.Integer => value is LUIWireValue.Int,
                LUIExtensionValueKind.DoubleValue =>
                    value is LUIWireValue.Float ||
                    value is LUIWireValue.Int,
                _ => false,
            };

        public static LUIWireValue Normalize(
            this LUIExtensionValueKind kind, LUIWireValue value) =>
            kind == LUIExtensionValueKind.DoubleValue &&
            value is LUIWireValue.Int number
                ? new LUIWireValue.Float(number.Value)
                : value;
    }

    public sealed class LUIExtensionProperty
    {
        public LUIExtensionProperty(
            string name, LUIExtensionValueKind kind, bool isRequired = false,
            LUIWireValue? defaultValue = null)
        {
            Name = name;
            Kind = kind;
            IsRequired = isRequired;
            DefaultValue = defaultValue;
        }

        public string Name { get; }
        public LUIExtensionValueKind Kind { get; }
        public bool IsRequired { get; }
        public LUIWireValue? DefaultValue { get; }
    }

    public sealed class LUIExtensionEventField
    {
        public LUIExtensionEventField(
            string name, LUIExtensionValueKind kind, bool isRequired = false)
        {
            Name = name;
            Kind = kind;
            IsRequired = isRequired;
        }

        public string Name { get; }
        public LUIExtensionValueKind Kind { get; }
        public bool IsRequired { get; }
    }

    public sealed class LUIExtensionEventSchema
    {
        public LUIExtensionEventSchema(
            string name = "", IReadOnlyList<LUIExtensionEventField>? fields = null)
        {
            Name = name;
            Fields = fields ?? (IReadOnlyList<LUIExtensionEventField>)
                System.Array.Empty<LUIExtensionEventField>();
        }

        public string Name { get; }
        public IReadOnlyList<LUIExtensionEventField> Fields { get; }
    }

    public sealed class LUIExtensionSpec
    {
        public LUIExtensionSpec(
            string identifier, string fingerprint,
            bool acceptsStandardChildren = false,
            IReadOnlyList<string>? childIdentifiers = null,
            IReadOnlyList<LUIExtensionProperty>? properties = null,
            IReadOnlyList<LUIExtensionEventSchema>? events = null)
        {
            Identifier = identifier;
            Fingerprint = fingerprint;
            AcceptsStandardChildren = acceptsStandardChildren;
            ChildIdentifiers = childIdentifiers ??
                (IReadOnlyList<string>)System.Array.Empty<string>();
            Properties = properties ??
                (IReadOnlyList<LUIExtensionProperty>)
                System.Array.Empty<LUIExtensionProperty>();
            Events = events ??
                (IReadOnlyList<LUIExtensionEventSchema>)
                System.Array.Empty<LUIExtensionEventSchema>();
        }

        internal static LUIExtensionSpec Tweak(
            string identifier, string fingerprint,
            IReadOnlyList<LUIExtensionProperty>? properties) =>
            new LUIExtensionSpec(
                identifier, fingerprint,
                acceptsStandardChildren: true,
                properties: properties)
            { IsTweak = true };

        public string Identifier { get; }
        public string Fingerprint { get; }
        public bool AcceptsStandardChildren { get; }
        public IReadOnlyList<string> ChildIdentifiers { get; }
        public IReadOnlyList<LUIExtensionProperty> Properties { get; }
        public IReadOnlyList<LUIExtensionEventSchema> Events { get; }
        public bool IsTweak { get; private set; }

        public LUIExtensionEventSchema? Event(string name)
        {
            foreach (LUIExtensionEventSchema candidate in Events)
            {
                if (candidate.Name == name) return candidate;
            }
            return null;
        }

        public LUIExtensionProperty? Property(string name)
        {
            foreach (LUIExtensionProperty candidate in Properties)
            {
                if (candidate.Name == name) return candidate;
            }
            return null;
        }
    }

    public sealed class LUIExtensionRegistry
    {
        static readonly Regex NamePattern =
            new Regex("^[a-z0-9]+(?:-[a-z0-9]+)*$", RegexOptions.Compiled);

        readonly Dictionary<string, LUIExtensionSpec> _registrations =
            new Dictionary<string, LUIExtensionSpec>();

        public bool IsFrozen { get; private set; }

        public static bool ValidExtensionName(string value) =>
            NamePattern.IsMatch(value);

        public void Register(LUIExtensionSpec registration)
        {
            if (IsFrozen)
            {
                throw new LUIBackendException("extension registry is frozen");
            }
            if (!ValidExtensionName(registration.Identifier))
            {
                throw new LUIBackendException("invalid extension identifier");
            }
            if (LUIWireSchema.IsStandardNodeName(registration.Identifier))
            {
                throw new LUIBackendException(
                    "extension identifier shadows a standard node");
            }
            if (_registrations.ContainsKey(registration.Identifier))
            {
                throw new LUIBackendException(
                    "extension identifier is already registered");
            }
            foreach (string child in registration.ChildIdentifiers)
            {
                if (!ValidExtensionName(child))
                {
                    throw new LUIBackendException(
                        "invalid extension child identifier");
                }
            }
            ValidateUniqueNames(registration.ChildIdentifiers, "child identifier");
            ValidateUniqueNames(
                Names(registration.Properties, property => property.Name),
                "property");
            ValidateUniqueNames(
                Names(registration.Events, evt => evt.Name), "event");
            foreach (LUIExtensionProperty property in registration.Properties)
            {
                if (!ValidExtensionName(property.Name))
                {
                    throw new LUIBackendException(
                        "invalid extension property name");
                }
                if (property.DefaultValue != null &&
                    !property.Kind.Accepts(property.DefaultValue))
                {
                    throw new LUIBackendException(
                        "invalid extension property default");
                }
            }
            foreach (LUIExtensionEventSchema evt in registration.Events)
            {
                if (!ValidExtensionName(evt.Name))
                {
                    throw new LUIBackendException(
                        "invalid extension event name");
                }
                ValidateUniqueNames(
                    Names(evt.Fields, field => field.Name), "event field");
                foreach (LUIExtensionEventField field in evt.Fields)
                {
                    if (!ValidExtensionName(field.Name))
                    {
                        throw new LUIBackendException(
                            "invalid extension event field name");
                    }
                }
            }
            _registrations[registration.Identifier] = registration;
        }

        public void RegisterTweak(
            string identifier, string fingerprint,
            IReadOnlyList<LUIExtensionProperty>? properties = null)
        {
            Register(LUIExtensionSpec.Tweak(identifier, fingerprint, properties));
        }

        internal void Freeze()
        {
            if (IsFrozen) return;
            foreach (LUIExtensionSpec registration in _registrations.Values)
            {
                foreach (string child in registration.ChildIdentifiers)
                {
                    if (!_registrations.ContainsKey(child))
                    {
                        throw new LUIBackendException(
                            "unknown extension child schema");
                    }
                }
            }
            IsFrozen = true;
        }

        public LUIExtensionSpec? Registration(string identifier) =>
            _registrations.TryGetValue(identifier, out LUIExtensionSpec? spec)
                ? spec
                : null;

        static IEnumerable<string> Names<T>(
            IEnumerable<T> items, System.Func<T, string> selector)
        {
            foreach (T item in items) yield return selector(item);
        }

        static void ValidateUniqueNames(IEnumerable<string> names, string label)
        {
            var seen = new HashSet<string>();
            foreach (string name in names)
            {
                if (!seen.Add(name))
                {
                    throw new LUIBackendException(
                        $"duplicate extension {label}");
                }
            }
        }
    }
}

File: platform/qt/lib/lui_extension_registry.cpp
#include "lui_extension_registry.h"

#include "lui_wire_schema.h"

#include <QSet>
#include <cmath>

namespace LUI {

namespace {

bool stringValue(const QVariant &v) { return v.typeId() == QMetaType::QString; }
bool boolValue(const QVariant &v) { return v.typeId() == QMetaType::Bool; }
bool intValue(const QVariant &v) {
  if (v.typeId() == QMetaType::Int) return true;
  if (v.typeId() != QMetaType::Double && v.typeId() != QMetaType::Float &&
      v.typeId() != QMetaType::LongLong &&
      v.typeId() != QMetaType::ULongLong) {
    return false;
  }
  const double n = v.toDouble();
  return std::isfinite(n) && std::floor(n) == n;
}
bool doubleValue(const QVariant &v) {
  switch (v.typeId()) {
  case QMetaType::Double:
  case QMetaType::Float:
  case QMetaType::Int:
  case QMetaType::LongLong:
  case QMetaType::ULongLong:
    return std::isfinite(v.toDouble());
  default:
    return false;
  }
}

} // namespace

bool ExtensionProperty::accepts(const QVariant &value) const {
  switch (kind) {
  case ExtensionValueKind::String:
    return stringValue(value);
  case ExtensionValueKind::Boolean:
    return boolValue(value);
  case ExtensionValueKind::Integer:
    return intValue(value);
  case ExtensionValueKind::Double:
    return doubleValue(value);
  }
  return false;
}

QVariant ExtensionProperty::normalize(const QVariant &value) const {
  if (!isRequired && !value.isValid()) return defaultValue;
  return value;
}

const ExtensionProperty *ExtensionSpec::property(const QString &name) const {
  for (const ExtensionProperty &prop : properties) {
    if (prop.name == name) return &prop;
  }
  return nullptr;
}

const ExtensionEventSchema *ExtensionSpec::event(const QString &name) const {
  for (const ExtensionEventSchema &schema : events) {
    if (schema.name == name) return &schema;
  }
  return nullptr;
}

bool ExtensionRegistry::validName(const QString &value) {
  if (value.isEmpty()) return false;
  bool lastWasDash = true; // first char cannot be a dash
  for (const QChar c : value) {
    const bool lower = (c >= u'a' && c <= u'z');
    const bool digit = (c >= u'0' && c <= u'9');
    if (lower || digit) {
      lastWasDash = false;
      continue;
    }
    if (c == u'-' && !lastWasDash) {
      lastWasDash = true;
      continue;
    }
    return false;
  }
  return !lastWasDash; // trailing dash rejected
}

bool ExtensionRegistry::add(const ExtensionSpec &spec, QString *error) {
  ExtensionSpec copy = spec;
  copy.isTweak = false;
  if (!validName(copy.identifier)) {
    *error = QStringLiteral("Invalid extension identifier: %1").arg(copy.identifier);
    return false;
  }
  if (standardNodeName(qPrintable(copy.identifier))) {
    *error = QStringLiteral(
        "Extension identifier %1 shadows a standard node")
        .arg(copy.identifier);
    return false;
  }
  if (copy.fingerprint.isEmpty() || copy.componentSource.isEmpty()) {
    *error = QStringLiteral(
        "Extension %1 requires a fingerprint and componentSource")
        .arg(copy.identifier);
    return false;
  }
  if (copy.isTweak && copy.childIdentifiers.size() != 1) {
    *error = QStringLiteral(
        "Tweak component %1 must list exactly one child identifier")
        .arg(copy.identifier);
    return false;
  }
  if (copy.isTweak && copy.acceptsStandardChildren) {
    *error = QStringLiteral(
        "Tweak component %1 cannot accept standard children")
        .arg(copy.identifier);
    return false;
  }
  QSet<QString> propNames;
  for (const ExtensionProperty &prop : copy.properties) {
    if (prop.name.isEmpty() || propNames.contains(prop.name)) {
      *error = QStringLiteral(
          "Duplicate or empty extension property name in %1")
          .arg(copy.identifier);
      return false;
    }
    propNames.insert(prop.name);
  }
  QSet<QString> eventNames;
  for (const ExtensionEventSchema &event : copy.events) {
    if (!validName(event.name) || eventNames.contains(event.name)) {
      *error = QStringLiteral(
          "Duplicate or invalid extension event name in %1")
          .arg(copy.identifier);
      return false;
    }
    eventNames.insert(event.name);
    QSet<QString> fieldNames;
    for (const ExtensionEventField &field : event.fields) {
      if (field.name.isEmpty() || fieldNames.contains(field.name)) {
        *error = QStringLiteral(
            "Duplicate or empty extension event field in %1.%2")
            .arg(copy.identifier, event.name);
        return false;
      }
      fieldNames.insert(field.name);
    }
  }
  if (m_frozen) {
    const ExtensionSpec *found =
        m_registrations.contains(copy.identifier)
            ? &m_registrations[copy.identifier]
            : nullptr;
    if (found == nullptr || found->fingerprint != copy.fingerprint) {
      *error = QStringLiteral(
          "Cannot register extension %1 after the registry was frozen")
          .arg(copy.identifier);
      return false;
    }
    return true;
  }
  if (m_registrations.contains(copy.identifier)) {
    const ExtensionSpec &existing = m_registrations[copy.identifier];
    if (existing.fingerprint != copy.fingerprint) {
      *error = QStringLiteral(
          "Extension %1 already registered with a different fingerprint")
          .arg(copy.identifier);
      return false;
    }
    // Same fingerprint → idempotent re-registration.
    return true;
  }
  for (auto it = m_registrations.constBegin(); it != m_registrations.constEnd();
       ++it) {
    if (it->fingerprint == copy.fingerprint) {
      *error = QStringLiteral(
          "Extension fingerprint collision between %1 and %2")
          .arg(copy.identifier, it->identifier);
      return false;
    }
  }
  m_registrations.insert(copy.identifier, copy);
  return true;
}

bool ExtensionRegistry::addTweak(const ExtensionSpec &spec, QString *error) {
  ExtensionSpec copy = spec;
  copy.isTweak = true;
  return add(copy, error);
}

const ExtensionSpec *ExtensionRegistry::registration(
    const QString &identifier) const {
  const auto it = m_registrations.constFind(identifier);
  return it == m_registrations.constEnd() ? nullptr : &it.value();
}

bool ExtensionRegistry::freeze(QString *error) {
  m_frozen = true;
  for (auto it = m_registrations.constBegin(); it != m_registrations.constEnd();
       ++it) {
    const ExtensionSpec &spec = it.value();
    if (spec.isTweak) {
      const QString &childId = spec.childIdentifiers.first();
      if (!m_registrations.contains(childId)) {
        *error = QStringLiteral(
            "Tweak component %1 requires child %2")
            .arg(spec.identifier, childId);
        return false;
      }
    }
    for (const QString &childId : spec.childIdentifiers) {
      if (childId == spec.identifier) {
        *error = QStringLiteral(
            "Extension component %1 cannot list itself as a child")
            .arg(spec.identifier);
        return false;
      }
      if (!m_registrations.contains(childId)) {
        *error = QStringLiteral(
            "Extension component %1 requires unknown child %2")
            .arg(spec.identifier, childId);
        return false;
      }
    }
  }
  return true;
}

} // namespace LUI
